docs/0000-parenting.md

```diff
@@ -0,0 +1,85 @@
+Imagine set of notes collecting quotes from favorite books:
+``` 
+Author Name
+Another Author Name
+Book Title 1
+Book Title 2
+Book Title 3
+Quote 1
+Quote 2
+Quote 3
+Quote 4
+Quote 5
+Quote 6
+``` 
+In flat notes list these notes are looking like a mess. With parenting they may look like this:
+``` 
+Author Name
+|-- Book Title 1
+    |-- Quote 1
+    |-- Quote 2
+|-- Book Title 2
+    |-- Quote 3
+Another Author Name
+|-- Book Title 3
+    |-- Quote 4
+    |-- Quote 5
+    |-- Quote 6
+``` 
+`Author Name` and `Another Author Name` are roots. There is smth like `parent: Author Name` in the metadata section of `Book Title 1` and `Book Title 2`. There is `parent: Book Title 1` in the metadata section of `Quote 1` and `Quote 2`. There is `parent: Book Title 2` in the metadata section of `Quote 3` etc.
+
+In the same time the 'root' notes `Author Name` and `Another Author Name` can be children of some other note(s) (say, `Reading`).
+
+It should be possible to add more than one parent to a note: `parent: Software, Programming`
+
+## Motivation
+
+Wiki links are great, but they are not representing the hierarchy of the notes. And when we are putting notes in foders/subfolders there is a need to create a special notes-placeholders, which exist to collect a backlinks. For example, if you keep your travel notes in `Travel` folder, there are good chances, there is a `Travel` note in this folder, which keeps all of the backlinks from all other travel-related notes.
+Another one drawback of using folders is that you can't have the same note in several folders without phisically copying it.
+Implementing parenting allows you to keep notes hierarchy clean and effective, while using wiki links for adding semantical context.
+
+## Detailed design
+
+The obvious way of implementing parenting is that the parent-relationship info could be stored in frontmatter metadata at the top of the file: `parent: Other document`. That could be made click-navigatable in the document, and we could build features on top of that:
+- Visualisation of the hierarchy
+- Children view (similar to backlinks explorer)
+- Whatever is needed
+
+There should be a way to ***hoist*** notes: being able to focus on/zoom into the specific note (using the specific note as the root of the hierarchy). 
+
+When you are in the middle of notes hierarchy (for example: `Author Name` > `Book Title 1`), the newly created notes should be created in the current context (`parent: Book Title 1`), until user explicitly specified another location.
+
+Drag and drop to change the parent notes?
+
+## Drawbacks
+
+Using parenting makes file explorer almost obsolete and may be confusing to the users, which are too tied to using files and folders.
+
+## Alternatives
+
+There may be a way to implement hierarchy in it's simplest form using just wiki links. For example, we may decide, that note, which has a link to other note is a child of the latter note. So the previous book quotes example may look like this:

```

Comment:
I would suggest that a way to annotate a link between two pages solves note hierarchy and also affords more complex linking between notes (e.g. multiple links, each one with different semantic meaning).

The markup could look something like this:

```markdown
[[some-other-page?Some context about why this link matters]]
```

I don't think that's a great way to do the markup, but it would work. An alternative would be:

```markdown
![[some-link]] Metadata goes here adjacent to the link
```

The `!` preceding the link on a newline would clue that this is an outbound link to the `some-link` page, and metadata goes adjacent to the link describing an edge.

Reply:
I prefer the first suggestion since ![[ ]] is already use for embedding a an image and even the text of the linked note (if not in Foam - not sure - then certainly in Obsidian, Markdown Memo, and others), and I don't think having the metadata outside of the link would work since links are often (generally) in-line.

Whatever the specific implementation, I think it will be crucial to have it be mapped to an actual field, rather than just be descriptive text - a field could be filtered and otherwise operated upon. Perhaps even multiple tag-like fields could be included within the metadata section

e.g. [[some-link?$relationship:parent;$compatible:true;$comment:some context about the link]]

Autocompletions within this would be even better!

Reply:
I saw the conflict with `![[]]` by chance after posting this.

Finding the right balance between technical rigor (multiple tag-like fields linking two notes) and ensuring the markdown is human friendly should be a first-order concern.

Dendron has an interesting approach to [references](https://www.dendron.so/notes/f1af56bb-db27-47ae-8406-61a98de6c78c.html) which would be more readable for complex/multi-property links than trying to inline it all in one link.

Reply:
Agreed - simplicity/human-friendly is key. Unorthodox file naming is why I was turned off of Dendron to begin with, but looking at it again, it is pretty impressive tool... I may have to explore some more.

Reply:
Checked it out. Some very cool stuff, but I don't like the highly rigid hierarchies. I've submitted an issue to ask if it is possible to ditch the "."-delimited hierarchies and use the front matter for parent/child relationships, but I suspect the answer will be no. 

That's another option here: don't put the link metadata in the actual links, integrate it with the front matter/link references/however Foam ends up handling links. You could just manually add 

```
title: xyz
date: 2020/08/26
parents: abc, def
children: 123, 789

---

Content
```

Since it would hopefully/need to be integrated with Foam's internal link management, perhaps when you enter a link [[note123]], there could be a popup with textfields for each attribute that allows for multiple auto-completed or comma separated entries .

Here's a similar example from TickTick, my task manager.

![image](https://user-images.githubusercontent.com/24255237/91353250-3e105080-e7a8-11ea-869e-88973543cd6d.png)

After you've added some, it shows them all as checkboxes when you go to edit again:

![image](https://user-images.githubusercontent.com/24255237/91353347-60a26980-e7a8-11ea-9fae-092304474bbd.png)